### PR TITLE
Add transduce

### PR DESCRIPTION
**Pull request by briancavalier**
Add `transduce` API for using composed transducers with most.js streams.

Still needs:

- [x] Unit tests
- [x] Docs

See #90

### CODE REVIEW

**Comment by briancavalier** [comment]
@jeffrose @kevinbeaty Code, tests, and docs are done!  I'd love to get your feedback before merging :)

**Comment by briancavalier** [comment]
Squashed commits for easier review

File: lib/combinator/transduce.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+TransduceSink.prototype.event = function(t, x) {
+	var next = this.xf.step(t, x);
+	if(next != null && next.__transducers_reduced__) {
+		this.sink.end(t, next.value);
+	}
+};
+
+TransduceSink.prototype.end = function(t /*, x */) {
+	this.xf.result(t);
```
Comment (kevinbeaty) [inline lib/combinator/transduce.js]: Is `x` here a possible additional item?  If so, it is valid to step one more time before calling result.

File: docs/api.md [hunk 2]
```diff
@@ -793,6 +795,37 @@ The `equals` function should accept two values and return truthy if the two valu
...
+## Transducer support
+
+### transduce
+
+####`stream.transduce(transducer) -> Stream`
+####`most.transduce(transducer, stream) -> Stream`
+
+Create a new stream by passing items through the provided transducer.
+
+[Transducers](http://simplectic.com/blog/2014/transducers-explained-1/) are composable transformations.  They may map, filter, add items to, drop items from, or otherwise transform an event stream.  The primary benefit of transducers is that they are composable and reusable across any data structures that support them (see note on performance below)
```
Comment (kevinbeaty) [inline docs/api.md]: Thanks for the shout out!

**Comment by kevinbeaty** [comment]
LGTM overall. Exciting stuff

File: docs/api.md (hunk 2 above)
Comment (briancavalier) [inline docs/api.md]: It's a great article :+1:

File: lib/combinator/transduce.js (hunk 1 above)
Comment (briancavalier) [inline lib/combinator/transduce.js]: Great question.  Most.js separates the notion of values "in the stream" and an "end signal value".  This is similar, but not identical, to who ES6 generates subtle separate `yield`ed values (which always have `done === true`) from a final `return`ed value (which has `done === false`).

So, now you've got me thinking that a better approach would be to wrap up `t` and `x` into a `{ time, value }` object and do `this.xf.result({ time: t, value: x })`, and then `Transformer.prototype.result`, could do `this.sink.end(time, value)`.  That would propagate this final value as the "end signal value".

**Comment by briancavalier** [comment]
Thanks @kevinbeaty, really appreciate the thorough review!

**Comment by kevinbeaty** [comment]
No problem.  

Just a heads up: The property that signals early termination `__transducers_reduced__` may be changing (for the better): cognitect-labs/transducers-js#20 . 

Not sure if that means you want to wait on this, or support the new version in the future.

**Comment by briancavalier** [comment]
Thanks for the heads up.  Any idea what the ETA is on the switchover to the new protocol?  If it's far enough out, I don't mind merging this asap and then updating when the new protocol lands.  If it's "close", then it may be better to wait.

**Comment by kevinbeaty** [comment]
Hard to say as there are multiple libraries involved. The discussion only started today but it seemed to reach a good common ground quickly. I don't think there is any harm in switching over later.  The current versions of the libraries follow the `__transducers_reduced__` convention that you implemented here.

**Comment by briancavalier** [comment]
Cool.  The only snag I can see on switching later is someone using an older transducer lib after most.js switches to support the newer protocol.  I think we can mitigate that via semver.  For example, I can release a new major (or minor if we're still < 1.0) version when most.js switches protocols.

So, yep, I'm cool with getting this merged asap and switching later.

**Comment by briancavalier** [comment]
Whew, ok, after a lot of thought and experimentation with different ways to thread *both* the event time and value through a transducer, I have it working.  Unfortunately, it required a stateful solution, but that's ok since it's hidden inside the machinery.  I also added a unit test for the scenario where a stream combinator ends the stream early (rather than a transducer combinator), to ensure that didn't wreak havoc on transducers at a later point in the stream.

I think this is ready to land!

**Comment by briancavalier** [comment]
Squashed again

**Comment by briancavalier** [comment]
Added jsdoc and squashed.  I'll merge this later today unless anyone has more suggestions.

**Comment by briancavalier** [comment]
:boom:

**Pull request merged**